File: spring/src/main/java/io/github/wavesync/entity/UserResonator.java
// MIGRATED to fastapi/app/models/user_resonator.py
package io.github.wavesync.entity;
import jakarta.persistence.*;
import lombok.*;
import java.util.ArrayList;
import java.util.List;




@Entity
@Table(name = "user_resonators")
@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
public class UserResonator {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "resonance_chain_level", nullable = false)
    private Integer resonanceChainLevel;

    @Column(name = "refine_level", nullable = false)
    private Integer refineLevel;

    @Builder.Default
    @Column(name = "is_deleted", nullable = false)
    private Boolean isDeleted = false;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "resonator_master_id", nullable = false)
    private ResonatorMaster resonatorMaster;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "weapon_master_id", nullable = false)
    private WeaponMaster weaponMaster;

    // finalStat 관계는 fastapi/app/models/user_resonator.py에도 양방향으로 미러링됨
    @OneToOne(mappedBy = "userResonator")
    private FinalStat finalStat;

    // userResonanceNodes 관계는 fastapi/app/models/user_resonator.py에도 양방향으로 미러링됨
    @Builder.Default
    @OneToMany(mappedBy = "userResonator")
    private List<UserResonanceNode> userResonanceNodes = new ArrayList<>();

    // userEchoes 관계는 fastapi/app/models/user_resonator.py에도 양방향으로 미러링됨
    @Builder.Default
    @OneToMany(mappedBy = "userResonator")
    private List<UserEcho> userEchoes = new ArrayList<>();
}

File: fastapi/app/models/user_resonator.py
from sqlalchemy import BigInteger, Boolean, Column, ForeignKey, Integer
from sqlalchemy.orm import relationship

from app.db.base import Base


class UserResonator(Base):
    """사용자가 등록한 공명자."""

    __tablename__ = "user_resonators"

    id = Column(BigInteger, primary_key=True)

    resonance_chain_level = Column(Integer, nullable=False)
    refine_level = Column(Integer, nullable=False)
    is_deleted = Column(Boolean, nullable=False, default=False)

    resonator_master_id = Column(BigInteger, ForeignKey("resonator_master.id"), nullable=False)
    weapon_master_id = Column(BigInteger, ForeignKey("weapon_master.id"), nullable=False)

    # 역방향(user_resonators)이 미사용이라 단방향으로만 연결한다.
    resonator_master = relationship("ResonatorMaster")
    weapon_master = relationship("WeaponMaster")

    final_stat = relationship("FinalStat", uselist=False, back_populates="user_resonator")
    user_resonance_nodes = relationship("UserResonanceNode", back_populates="user_resonator")
    user_echoes = relationship("UserEcho", back_populates="user_resonator")
